### PR TITLE
[fully_async] fix: replace routed_experts on partial rollout resume i…

### PR DESCRIPTION
**Pull request by NoonePauseferg**
### What does this PR do?

Fixes a bug in `FullyAsyncLLMServerManager.generate()` where `routed_experts` was incorrectly concatenated via `torch.cat` during partial rollout resume, causing duplicated routing data and broken MoE expert replay in the actor.

sglang returns `routed_experts` for the **full sequence** (prompt + all generated tokens). Evidence from sglang source:

1. **[`io_struct.py#L1020`](https://github.com/sgl-project/sglang/blob/v0.5.9/python/sglang/srt/managers/io_struct.py#L1020)** — field definition:
   ```python
   # The routed experts for each token, including both input and output tokens
   # routed_experts[i] is a tensor of shape (token, layer, top_k) for request i
   routed_experts: List[Optional[torch.Tensor]]
   ```

2. **`schedule_batch.py`** — `seqlen` used to collect routing covers the full sequence:
   ```python
   @property
   def seqlen(self) -> int:
       return len(self.origin_input_ids) + len(self.output_ids)
   ```

3. **`topk.py#L1049-1051`** — capture is unconditional (no prefill/decode check):
   ```python
   get_global_experts_capturer().capture(layer_id=layer_id, topk_ids=topk_ids)
   ```

4. **`scheduler_output_processor_mixin.py#L105-111`** — collection uses full `seqlen`:
   ```python
   req.routed_experts = get_global_experts_capturer().get_routed_experts(
       req_pool_idx=req.req_pool_idx,
       seqlen=req.seqlen,  # origin_input_ids + output_ids
       req_to_token_pool=self.req_to_token_pool,
   )
   ```

When partial rollout resumes after abort, the input becomes `prompt + already_generated_tokens`. sglang re-processes the entire input during prefill and returns `routed_experts` covering all positions. The old code concatenated this with the previous `routed_experts`:

```
old routing:    prompt + A B C
new routing:    prompt + A B C + D E
concat result:  prompt + A B C + prompt + A B C + D E   <-- duplicated!
expected:       prompt + A B C + D E
```

This shifted the routing and caused incorrect MoE expert replay, leading to `actor/ppo_kl` spikes.

**Fix:** replace `routed_experts` instead of concatenating, since the resumed call already covers all positions.

Related: #4348 (partial rollout RFC), #4101 (R3 router replay), #5344 (R3 in fully async)

### Checklist Before Starting

- [x] Search for similar PRs: https://github.com/verl-project/verl/pulls?q=routed_experts+partial_rollout
- [x] Format the PR title as `[{modules}] {type}: {description}`

### Test

- Ran async training with `partial_rollout=True` and `enable_rollout_routing_replay=True` (R3 mode)
- Verified `actor/ppo_kl` no longer spikes after partial rollout resume
- Verified `routed_experts` tensor shape matches `(prompt_len + response_len, num_layers, top_k)` after resume

### Design & Code Changes

Single-line change in `verl/experimental/fully_async_policy/agent_loop/agent_loop.py`:

```diff
- if output.routed_experts is not None:
-     if final_output.routed_experts is None:
-         final_output.routed_experts = output.routed_experts
-     else:
-         final_output.routed_experts = torch.cat([final_output.routed_experts, output.routed_experts], dim=0)
+ # sglang returns routed_experts for the full sequence (prompt + all tokens),
+ # so on partial rollout resume the new output already covers all positions.
+ if output.routed_experts is not None:
+     final_output.routed_experts = output.routed_experts
```

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: This is an async distributed training bug that requires multi-node sglang + megatron setup with MoE model and partial rollout enabled. Not feasible to reproduce in CI.
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1).

### CODE REVIEW

**Comment by ArronHZG** [comment]
Please use `pre-commit run --all-files` to format code.

File: verl/experimental/fully_async_policy/agent_loop/agent_loop.py [hunk 1]
```diff
@@ -96,11 +95,10 @@ async def generate(
             final_output.token_ids.extend(output.token_ids)
             if output.log_probs is not None:
                 final_output.log_probs.extend(output.log_probs)
+            # sglang returns routed_experts for the full sequence (prompt + all tokens),
+            # so on partial rollout resume the new output already covers all positions.
             if output.routed_experts is not None:
-                if final_output.routed_experts is None:
-                    final_output.routed_experts = output.routed_experts
-                else:
-                    final_output.routed_experts = torch.cat([final_output.routed_experts, output.routed_experts], dim=0)
```
Comment (wuxibin89) [inline verl/experimental/fully_async_policy/agent_loop/agent_loop.py:103]: I think we should extract routed_experts corresponding to  partially generated `output.token_ids` to reflect different model version?
```python
if len(output.token_ids) > 0:
  final_output.routed_experts = torch.cat([final_output.routed_experts, output.routed_experts[-len(output.token_ids):]], dim=0)
```

**Review by wuxibin89 — CHANGES_REQUESTED** [review]

File: verl/experimental/fully_async_policy/agent_loop/agent_loop.py (hunk 1 above)
Comment (NoonePauseferg) [reply verl/experimental/fully_async_policy/agent_loop/agent_loop.py:103]: Good idea at first glance, but I think simple replace is actually better here.

The actor update always runs with the latest weights (v2). If we replay v1's routing with v2's weights, the recomputed logprob is neither pi_v1 nor pi_v2 — it's an inconsistent mix. Simple replace gives v2 routing + v2 weights = clean pi_v2, which makes the importance ratio well-defined.

The codebase has a rollout correction pipeline (IS weights, rejection sampling) that can mitigate staleness when enabled, but it's off by default (`bypass_mode=True`).

That said, worth experimenting with both approaches — the per-iteration routing might still help in practice if routing signal matters more than weight consistency.

### TODO: experiments (64-expert 10B MoE model)
- [x] Compare simple replace vs per-iteration slicing with partial rollout enabled
- [x] Track `rollout_corr/kl`, `ppo_kl`, and reward curves for both approaches
- [ ] Test with `bypass_mode=False` + rollout IS enabled to see if correction compensates for either approach

Comment (wuxibin89) [reply verl/experimental/fully_async_policy/agent_loop/agent_loop.py:103]: Could you help to do some comparison experiment?

Comment (NoonePauseferg) [reply verl/experimental/fully_async_policy/agent_loop/agent_loop.py:103]: yeah - im in process

Comment (wuxibin89) [reply verl/experimental/fully_async_policy/agent_loop/agent_loop.py:103]: @NoonePauseferg Merge this PR to use latest router_experts as baseline. Please update per-iteration slicing experiment result if there any update.

**Review by wuxibin89 — APPROVED** [review]

File: verl/experimental/fully_async_policy/agent_loop/agent_loop.py (hunk 1 above)
Comment (NoonePauseferg) [reply verl/experimental/fully_async_policy/agent_loop/agent_loop.py:103]: recap_vs_simple
grad_norms


i have done comparison experiments on megatron_sglang setup
- version where we remember routing show more stable training results

Comment (NoonePauseferg) [reply verl/experimental/fully_async_policy/agent_loop/agent_loop.py:103]: created following pr here: https://github.com/verl-project/verl/pull/6046
